CAVEATS.md

```diff
@@ -0,0 +1,175 @@
+when the kernel is compiled with `clang`. We recommend `gcc` for working with
+SMP configurations of seL4.
+
+The combination of SMP and hypervisor extensions is supported and should be
+generally stable, but like the plain SMP configuration it is not formally
+verified.
+
+The combination of SMP and MCS is supported and is receiving active development,
+but it is less explored and less tested. It should still be considered
+experimental. There are no supported Armv7-a boards for SMP+MCS, only Armv8-a,
+RISC-V, and Intel. It is tested with `gcc` on `hifive`, `tqma8xqp1gb`,
+`odroidc4`, `zynqmp`, `tx1`, `tx2`, `pc99-32`, and `pc99-64`.
+
+The combination of SMP, MCS, and hypervisor extensions is currently supported on
+AArch64 only. It is less tested with lower code coverage; currently with `gcc`
+only, on `odroidc4`, `tx1`, and `tx2`.
+
+The combination of SMP and domain scheduler is not supported. The SMP
+configuration is not expected to satisfy strong intransitive non-interference
+for information flow.
+
+See the [seL4 issue tracker][issues] and the [sel4test issue tracker][sel4test
+issues] for details using the labels `MCS` and `SMP` for finding issues on these
+configurations.
+
+As these are unverified configurations, standard C implementation defects are
+possible and not excluded as in verified seL4 configurations.
+
+Supporting a static multi-kernel configuration with formal verification is on
+the roadmap for the AArch64 architecture, with initial work begun. We expect
+multi-kernel configurations to be more robust than SMP configurations, because
+they are simpler and closer to the current sequential seL4 proofs.
+
+In a multi-kernel configuration, each CPU core runs a separate instance of seL4,
+with each kernel instance getting access to disjoint subsets of memory of the
+machine. User-level memory can be shared as device-untyped memory, which the
+kernel manages but does not access. These configurations can already be set up
+without kernel changes by providing suitable device tree overlays to each kernel
+instance. Further work is planned to make such configurations easier to use and
+more robust against unsafe use/configurations, e.g. by managing IRQ controller
+access for each instance.
+
+## Re-using Address Spaces
+
+Before an ASID/page directory/page table can be reused, all frame caps
+installed in it should be revoked. The kernel will not do this automatically
+for the user.
+
+If, for instance, page cap `c` is installed in the address space denoted by a
+page directory under ASID `A`, and the page directory is subsequently revoked or
+deleted, and then a new page directory is installed under that same ASID `A`,
+the page cap `c` will still retain some authority in the new page directory,
+even though the user intention might be to run the new page directory under a
+new security context. The authority retained is to perform the unmap operation
+on the page the cap `c` refers to.
```

Comment:
Surely this only applies to platforms that can't invalidate all TLB entries by ASID? ARM invalidates the ASID when unmapping the page table or destroying a mapped vspace. A quick look at x86's and riscv's `deleteASID()` confirms the same behaviour there. So I'm mystified how the stale mappings can be still active after a delete.

Does it only apply to re-using ASIDs and page tables without a delete and assigning it to a different TCB? Perhaps say that explicitly? But why would the user expect the kernel to unmap existing mappings from a page table automatically for any reason?

Revoking a page table does nothing, as there are no derived objects, so why mention it?

Reply:
Maybe it's clearer if you say that page caps store the vspace's ASID when mapped and will keep granting rights to do operations on that mapped virtual address range with that ASID even after that ASID has been re-used for another vspace.

Reply:
I can give that a shot

Reply:
Does the new version work better?